File: LeetCode/1.py
# Own
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        for i in range(len(nums)):
                for j in range(len(nums)):
                    if j != i and nums[j] + nums[i] == target:
                        return [j, i]
                    
# Optimal
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        hash_table = {}
        for i in range(len(nums)):
            complement = target - nums[i]
                
            if complement not in hash_table:
                hash_table.update({nums[i]: i})
            else:    
                return [i, hash_table[complement]]
            

File: leetcode/1.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int n = nums.length;
        HashMap<Integer, Integer> numMap = new HashMap<>();

        for (int i = 0; i < n; ++i) {
            int num = nums[i];
            if (numMap.containsKey(target - num)) {
                return new int[] { numMap.get(target - num), i };
            }
            numMap.put(num, i);
        }
        return new int[0];
    }
}
